test/gtest/layout_transpose.cpp

```diff
@@ -0,0 +1,566 @@
+        return;
+
+    auto idx = miopen::mismatch_idx(t_tst.data, t_ref.data, compare_equal<T>);
+
+    EXPECT_GE(idx, miopen::range_distance(t_ref))
+        << "diff at:" << idx << ", " << _tst << ":" << t_tst[idx] << ", " << _ref << ":" << t_ref[idx];
+}
+
+struct transpose_dims
+{
+    static constexpr uint32_t image_size_rand_offset {29};
+    static constexpr uint32_t image_size_rand_range {13};
+    static uint32_t get_max_image_size()
+    {
+        return image_size_rand_offset + image_size_rand_range - 1;
+    }
+    static std::vector<uint32_t> get_image_size()
+    {
+        std::vector<uint32_t> v = {1, 9, 14};
+        v.push_back(prng::gen_off_range(image_size_rand_offset, image_size_rand_range));
+        return v;
+    }
+    static std::vector<uint32_t> get_channel_size()
+    {
+        std::vector<uint32_t> v = {3, 11};
+        v.push_back(prng::gen_off_range(19, 7));
+        return v;
+    }
+    static std::vector<uint32_t> get_batch_size()
+    {
+        std::vector<uint32_t> v = {1, 2, 4};
+        v.push_back(prng::gen_off_range(3, 4));
+        return v;
+    }
+};
+
+struct transpose_invoke_param : public miopen::InvokeParams
+{
+    ConstData_t src = nullptr;
+    Data_t dst      = nullptr;
+
+    transpose_invoke_param(ConstData_t src_, Data_t dst_) : src(src_), dst(dst_) {}
+    transpose_invoke_param(miopen::InvokeType type_, ConstData_t src_, Data_t dst_)
+        : InvokeParams{type_}, src(src_), dst(dst_)
+    {
+    }
+
+    Data_t GetWorkspace() const { return nullptr; }
+    std::size_t GetWorkspaceSize() const { return 0; }
+};
+
+} // namespace batched_transpose
+
+using namespace batched_transpose;
+
+template <typename T, class TRANSPOSE_SOL>
+struct LayoutTransposeTest_2D
+    : public ::testing::TestWithParam<
+          std::tuple<uint32_t, uint32_t>>
+{
+protected:
+    miopen::ExecutionContext    ctx;
+
+    uint32_t n;
+    uint32_t c;
+
+    virtual void SetUp() override
+    {
+        std::tie(n, c) = GetParam();
+    }
+    void RunTest()
+    {
+        auto&& handle = get_handle();
+
+        auto wh             = transpose_dims::get_max_image_size();
+        auto max_tensor_sz  = n * c * wh * wh;
+        auto dst_dev        = handle.Create(sizeof(T) * max_tensor_sz);
+
+        auto H = transpose_dims::get_image_size();
+        auto W = transpose_dims::get_image_size();
+
+        for (auto h : H)
+        {
+            for (auto w : W)
+            {
+                std::vector<int> tensor_len = {
+                    static_cast<int>(n),
+                    static_cast<int>(c),
+                    static_cast<int>(h),
+                    static_cast<int>(w)
+                };
+
+                std::vector<int> tensor_strides;
+
+                std::string layout_default = miopen::tensor_layout_get_default(4);
+                std::string layout_string  = tensor_layout_to_string(miopen::tensor_layout_nchw);
+
+                miopen::tensor_layout_to_strides(
```

Comment:
the body of the loop from here on is similar to the body of the loop in the `3D test`. Please factor out the common code into helper functions.

Reply:
I considered that as well and decided the benefit does not justify the cost. It is very important that this PR gets merged before the end of the sprint. If you would like to discuss further refactoring, let's do that next sprint and do any new work under another ticket.

Reply:
let's do it in the next sprint. You should create a new ticket.

Reply:
I should have added that those helper functions will probably be useful in several other unit test files. Doing it the right way is far from a trivial exercise.